Migrate from JUnit 4 to JUnit 5

sijie123 opened the pull request:
Our tests use JUnit4. 

JUnit 5 provides many APIs that simplify our tests. For example, instead of a lengthy try/catch procedure to test for correct exception handling behaviour, we can now use assertThrows. This makes it easier for readers to understand our code, and to implement more tests.

Let's upgrade to use JUnit 5.

Resolves #874 

Progress/To-do:
[x] Change imports to JUnit5 version. (e.g. org.junit.jupiter.api.Test instead of org.junit.Test)
[x] Remove Test being used for specifying expectations. We use this to specify timeout in PersonListPanelTest. Instead we should now use assertTimeoutPreemptively. (Done in #863)
[x] Rename @Before to @BeforeEach
[x] Rename @After to @AfterEach
[x] Rename @BeforeClass to @BeforeAll
[x] Rename @AfterClass to @AfterAll
[-] Group assertions together using assertAll for related assertions (Not sure if we want to do this in this PR? Leaving it as separate asserts right now as feedback to user is clearer when tests fail)
[x] Remove @Rule and @ClassRule. Instead, we need to @ExtendWith, which generally requires creating an extension class.

pyokagan commented:
@sijie123 Can we put everything into one single "migrate to JUnit5" PR if possible?

Long ago, we introduced a `history` command feature that was implemented with the expectation that its code could be shared with the upcoming undo/redo feature.

Turns out, after the `history` command PR was merged, that it couldn't share the same mechanism as the undo/redo feature.

Lesson: We never know if a commit is actually required in accomplishing a goal, unless we *actually* accomplish that goal :-)

sijie123 commented:
> @sijie123 Can we put everything into one single "migrate to JUnit5" PR if possible?

Sure. In that case, I'll make this PR the gigantic one then.
Topic changed.

Zhiyuan-Amos commented on src/test/java/seedu/address/testutil/Assert.java:
```diff
@@ -1,23 +1,43 @@
+     *
+     * @deprecated use {@see assertThrowsWithMessage(Class<? extends Throwable>, String, Executable)} instead.
      */
+    @Deprecated
```
We don't use `deprecated` because we are going to replace all usages of this API with the newer one and remove this method from the codebase :P

sijie123 replied:
Yup. Will remove the old version once I finish migrating all the API calls. Then I will rebase the Assert.java on top.

Zhiyuan-Amos commented on src/test/java/seedu/address/testutil/Assert.java:
```diff
@@ -1,23 +1,43 @@
      */
-    public static void assertThrows(Class<? extends Throwable> expected, VoidCallable callable) {
-        assertThrows(expected, null, callable);
+    public static void assertThrows(Class<? extends Throwable> expectedType, Executable executable,
```
Sorry, I think I wasn't clear in my earlier reply. What I meant was: We just need:

1. `assertThrows(Class<? extends Throwable> expectedType, Executable executable)` which you have implemented.
1. `assertThrows(Class<? extends Throwable> expectedType, Executable executable, String expectedMessage)`

There's no need for this method because we don't use it; the purpose of this PR is to simplify existing code upon upgrading the JUnit version.

sijie123 replied:
We don't need (2)? I thought it's quite commonly used in AB4.
For example:
https://github.com/se-edu/addressbook-level4/blob/921e61f831cebaa4a378c4ce9c59aee2f54e0948/src/test/java/seedu/address/logic/commands/CommandTestUtil.java#L108-L109

I was thinking of putting them all in one place under Assert.java

Zhiyuan-Amos replied:
We need (2), but we don't need this :P

What this method is doing is what you have said earlier:

> `message` is the message to be displayed when the test fails.

Which we don't use in our existing codebase.

sijie123 replied:
Oh. I see. I saw some code that used it, but it's okay, I'll probably just write a function only in the one or two places that use it.
Or perhaps just remove the message out right.

sijie123 commented:
I'm ready for the onslaught... or maybe not :P

Jokes aside, I would need help with organising my commits. I've tried to separate them as much as possible, although it's still not very ideal. For example, adding support for TempDir relies on using JUnit 5 Runner API, and JUnit 5 runner API won't support the old TemporaryFolder rule, so I'm not sure how to split the concerns in this case.

Zhiyuan-Amos commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -70,7 +66,7 @@ public void execute_commandExecutionError_throwsCommandException() {
 
     @Test
-    public void execute_validCommand_success() {
+    public void execute_validCommand_success() throws Exception {
```
the additional `throws Exception` is unclear.

sijie123 commented:
Thanks @pyokagan and @Zhiyuan-Amos for your reviews.
I'm still trying to figure out how granular the commits should be for this project, and it seems that I've severely overestimated how granular I should go.

I'll start by splitting up the very first commit (Assert: Use JUnit 5 assertions). During this process, I will probably split up some parts which I find less relevant to the migration, into other separate PRs.

sijie123 commented:
Hi @pyokagan and @Zhiyuan-Amos 
I'm working on splitting the commits into much smaller granularity. I've been able to work out the first few commits (hopefully they're what is expected), but I'm kinda stuck at [v2 9/16](https://github.com/se-edu/addressbook-level4/pull/959/commits/0c84c78230f175526d9d82625bc6e920a90ea35b).
In this commit, I attempt to migrate the UI test stuff to become JUnit 5 compatible.

In JUnit 5, there's no longer support for extensions. So both TemporaryFolder and UiPartRule/StageRule need to go. I thought it'd be too large a commit to change both things at once, but I can't change one without breaking the other.
If I change UiPartRule --> UiPartExtension and StageRule --> StageExtension, then I have to use the new JUnit 5 test runner (org.junit.jupiter.api.Test) to register these extensions (otherwise they will have no effect!). But if I use the new test runner, then the TemporaryFolder rule will not work, and vice versa.

I'm not sure how to break this down into a smaller, "single responsibility" change. Can I get some advice on this?

Zhiyuan-Amos commented:
@sijie123 

> In JUnit 5, there's no longer support for extensions. So both TemporaryFolder and UiPartRule/StageRule need to go... I can't change one without breaking the other.

Didn't verify this, but if this is indeed true, then note it down in your commit message. Your commit message can look something like:

```
Currently, we use JUnit 4 test runner (org.junit.Test) to execute our unit and system testing.

As we migrate to JUnit 5 to adopt new features such as extensions and temporary directories, we need to use the new JUnit 5 test runner.
This allows us to use JUnit 5 features such as extensions model, pre- and post-test callbacks, as well as temporary directories.

However, JUnit 5 no longer supports UiPartRule, StageRule and TemporaryFolder rule

Let's update to use JUnit5 test runner and replace UiPartRule with..... etc etc
```

(I copied the 1st 2 paras from 10/16's commit message).

Zhiyuan-Amos commented on src/test/java/seedu/address/testutil/Assert.java:
```diff
@@ -1,45 +1,32 @@
-     * If there's no need for the verification of the exception's error message, call
-     * {@code assertThrows(Class<? extends Throwable>, VoidCallable)} instead.
-     * {@see assertThrows(Class<? extends Throwable>, VoidCallable}
+    /** Asserts that the {@code executable} throws the {@code expectedType} Exception with the
```
The comment should be on the next line

Zhiyuan-Amos replied:
The comment should be on the next line

Zhiyuan-Amos commented on src/test/java/seedu/address/testutil/Assert.java:
```diff
@@ -1,45 +1,32 @@
+    /** Asserts that the {@code executable} throws the {@code expectedType} Exception with the
+     * {@code expectedMessage} message. If there's no need for the verification of the exception's error
+     * message, call {@code assertThrows(Class<? extends Throwable>, Executable)} instead.
+     * {@see assertThrows(Class<? extends Throwable>, Executable}}
```
> {@see assertThrows(Class<? extends Throwable>, Executable}}

The 2nd last character should be a `)` instead of the curly brace.

Zhiyuan-Amos commented on src/test/java/seedu/address/testutil/Assert.java:
```diff
@@ -1,45 +1,32 @@
-     * {@code assertThrows(Class<? extends Throwable>, VoidCallable)} instead.
-     * {@see assertThrows(Class<? extends Throwable>, VoidCallable}
+    /** Asserts that the {@code executable} throws the {@code expectedType} Exception with the
+     * {@code expectedMessage} message. If there's no need for the verification of the exception's error
```
> {@code expectedMessage} ~~message~~.

sijie123 replied:
Fixed. Thanks.

Zhiyuan-Amos commented on src/test/java/seedu/address/testutil/Assert.java:
```diff
@@ -1,45 +1,32 @@
+     * Asserts that the {@code executable} throws the {@code expectedType} Exception.
+     * This is a wrapper method that invokes
+     * {@code Assertions.assertThrows(Class<? extends Throwable>, Executable)}, to maintain consistency
+     * with our custom {@see assertThrows(Class<? extends Throwable>, String, Executable)} method.
```
We should inform users not to call JUnit 5's native assertThrows method.

sijie123 replied:
Agreed. Added.

Zhiyuan-Amos commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -70,7 +70,7 @@ public void execute_commandExecutionError_throwsCommandException() {
 
     @Test
-    public void execute_validCommand_success() {
+    public void execute_validCommand_success() throws Exception {
```
do we need this?

sijie123 replied:
Yes.
AssertCommandSuccess throws a checked exception because it executes the command directly. Since logic.execute throws a checked exception, this check needs to be passed down the chain 

^Should I include this reason in the commit message?

Zhiyuan-Amos replied:
Nice, I missed that out. Don't have to include it in the commit message.

Zhiyuan-Amos commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -106,55 +106,47 @@ public void getFilteredPersonList_modifyList_throwsUnsupportedOperationException
-     * and also confirms that the following two parts of the LogicManager object's state are as expected:<br>
-     *      - the internal model manager data are same as those in the {@code expectedModel} <br>
-     *      - {@code expectedModel}'s address book was saved to the storage file.
+     * Executes the command, confirms that the exception is thrown and that the result message and model is correct.
```
the header comment is too brief (e.g. what does it mean that `the result message and model is correct`?)

Zhiyuan-Amos commented:
I'll stop here first. Do take a look at your commit messages and follow our commit messages convention [here](https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message):

```
{current situation} -- use present tense

{why it needs to change}

{what is being done about it} -- use imperative mood

{why it is done that way}

{any other relevant info}
```

Compare your [3/19] commit message with the comment I mentioned above https://github.com/se-edu/addressbook-level4/pull/959#pullrequestreview-215354903. :)

Otherwise, I think you've done a great job in tidying up the code base! This PR is huge, and I'm wondering whether we should split up this PR. You mentioned in your first post that: 

> This is the first of a series of (proper) PRs that will migrate AB4 to use JUnit 5

But your PR seems to be doing nearly everything :P

sijie123 commented:
> [1/19]
> 
> Can this commit be shifted to a later commit, where the next commit is that of migrating to JUnit's native TemporaryFolder?
> 
> > Currently, JUnit 5.1.0 is used to write JUnit 5 styled tests.
> 
> This paragraph can be more specific e.g to write JUnit 5 styled tests such as creating Temporary Folders.

Agreed. Makes sense to move this to a later commit. Reworded commit to add more relevant information.

> [2/19]
> 
> > Currently, checkstyle is configured for JUnit 4 test annotations.
> 
> There's no need for `Currently` because our commit message convention is such that the first paragraph is about `what's happening currently`.
> 
> > It allows exceptions in coding style in cases like **such as** public variable
> > declaration **declaration of non-static, non-final and mutable public variables**, if **these variables are** preceded by a Rule tag **specified annotation members**. which **This** is necessary for correct variable assignment and execution of unit tests.
> 
> > For example, BeforeClass is now BeforeAll.
> 
> I think this should be placed is [18/19] when you remove existing JUnit 4 checkstyle properties.
> 
> > This allows us to use JUnit 5 annotations while suppressing false
> > positives in CheckStyle.
> 
> This should be merged into the first paragraph as part of `necessary for correct variable assignment and execution of unit tests.`
> 
> > add TempDir and RegisterExtension annotation in addition to default
> > annotations [1] in VisibilityModifier module.
> 
> I think there's no need to talk about what're the `default annotations` because we are simply adding what we need (i.e. if it happens that we don't need a particular default annotation, then we shouldn't be adding it in).
> 
> It turns out that we don't use `@Rule` `@ClassRule` and `@VisibleForTesting`. Is there a reason for including them in?

I think my commit message might have given the wrong impression. I will reword it.
I meant to say that currently our checkstyles only allow for "poor coding practices" for tests written using JUnit 4 syntax. For example, when you use @Rule before you declare a public variable, checkstyle will not complain. However, we need to apply this same exceptions to JUnit 5 syntax, otherwise code such as  
```
@TempDir 
public Path tempDir;
```
will get flagged by CheckStyle.

It's not possible to put this at [18/19], because commits like migrating to use TempDir will not pass the checkstyle. Hence, I have to put this upfront. That said, I will move it a little further back to introduce it only when its absolutely necessary.
For the same reason, I cannot immediately discard our exceptions meant for JUnit 4 syntax, because when this commit is introduced, we haven't gotten rid of all JUnit 4 code either.

The @Rule, @ClassRule @VisibilityForTesting stuff is there by default. However, if we override the value to include Junit 5's TempDir and RegisterExtension, the Rule,ClassRule,VisibleForTesting will not be included by default. To maintain compatibility with existing JUnit 4 code, I have to add the whole list.

I will reword the commit to reflect this thought process.



> [3/19]
> 
> > Currently, Assert implements our custom assertThrows using try/catch.
> 
> ```
> Assert contains a custom assertThrows method. 
> 
> This is additional, unnecessary code to maintain as JUnit 5's Assertions.assertThrows performs the same functionality Assert#assertThrows.
> 
> Let's replace Assert#assertThrows with JUnit 5's assertThrows.
> ```

Hmm. This is nice and short. I think I tend to err on the side of verbosity, cuz I'm still not too sure how much I should justify things. 



> [7/19]
> 
> > Currently, in our tests, we use a mix of 4 different ways to test
> > exception throwing. We use:
> > 
> > * try/catch - catch and ensure we receive the correct exception.
> 
> In that case, should [8/19] be in this commit too? If it's in, then I think this commit is kinda too big :P

I intended to split up 7 and 8 because 8 involved significantly more things within IndexTest. There's the idea of removing useless tests within IndexTest, but it only makes sense because we use the Assert#assertThrows API.
It didn't make sense for me to migrate IndexTest#assertCreateFailure together in [7/19], then in [8/19] remove createOneBasedIndex and createZeroBasedIndex. If I had done this, I would have needed if/else statements in IndexTest#assertCreateFailure, which defeats the purpose of using assertZeroBasedFailure and assertOneBasedFailure. Then the commit by itself doesn't make sense - we essentially end up combining assertZero- and assertOne- BasedFailure by calling assertCreateFailure, then we use if/else inside assertCreateFailure to split up the cases.

That's why I split up 7 and 8, because I think the changes in [8/19] makes sense / can be well justified only if its done entirely in one-shot. 

> [9/19]
> 
> > Notes: this commit only serves to introduce the DirectoryInitUtil
> > extension class. As extensions can only be used in JUnit 5, migration
> > to have tests use this DirectoryInitUtil extension will only be done
> > in a later commit. This is so as to ensure that every commit in the
> > meantime remains buildable.
> 
> I think this paragraph is no longer relevant because we have already migrated to 5.4.0 in the first commit?

Thanks for the catch . I meant to say that tests need to first use the JUnit 5 runner* before they can use the TempDir API and consequently the DirectoryInitUtil extension. The migration to JUnit 5.4.0 only allows existing classes already on the JUnit 5 runner to use the TempDir API. I will reword this commit.

Zhiyuan-Amos commented:
> For example, BeforeClass is now BeforeAll.

> > I think this should be placed is [18/19] when you remove existing JUnit 4 checkstyle properties.

> > > It's not possible to put this at [18/19], because commits like migrating to use TempDir will not pass the checkstyle. Hence, I have to put this upfront. That said, I will move it a little further back to introduce it only when its absolutely necessary.

What I meant was that the *commit message* should be placed in [18/19]. In [18/19], you mentioned that:

> After migrating all tests to JUnit 5, we no longer need these exceptions for annotations specific to JUnit 4.

So, you could be a bit clearer by specifying what has been replaced by what (e.g. `BeforeClass` replaced by `BeforeAll`), which explains why we are removing certain annotations. In [2/19], you've added annotations used by JUnit 5 (`BeforeAll`), but you haven't removed the ones used by JUnit 4, so it's not necessary to make the comparison of `BeforeClass` being replaced by `BeforeAll`. Sorry that I didn't explain myself clearly.

> It allows exceptions in coding style in cases like such as public variable
declaration declaration of non-static, non-final and mutable public variables, if these variables are preceded by a Rule tag specified annotation members. which This is necessary for correct variable assignment and execution of unit tests.

Yup, this can stay in [2/19].

> The @rule, @ClassRule @VisibilityForTesting stuff is there by default. However, if we override the value to include Junit 5's TempDir and RegisterExtension, the Rule,ClassRule,VisibleForTesting will not be included by default. To maintain compatibility with existing JUnit 4 code, I have to add the whole list.

Yup apologies for this, I missed out that you removed things such as `@Rule` in the later commit.

> I intended to split up 7 and 8 because 8 involved significantly more things within IndexTest. There's the idea of removing useless tests within IndexTest, but it only makes sense because we use the Assert#assertThrows API.

Ok. Do note that you removed `StringUtilTest#assertExceptionThrown(...)` as well, so, should that be placed in another commit too?

It'll be helpful if you mention in your commit message that [7/19] will only be performing mechanical changes, while the subsequent commit(s) (e.g. [8/19]) will perform more complex changes. This helps reviewers like myself understand why you're splitting it up :)

> Hmm. This is nice and short. I think I tend to err on the side of verbosity, cuz I'm still not too sure how much I should justify things.

You're doing great for a first-timer, keep up the great work! :)

sijie123 commented:
> What I meant was that the _commit message_ should be placed in [18/19]. In [18/19], you mentioned that:
Agreed. Reading my commit message again, I realised [18/19] was rather unclear. 


> Ok. Do note that you removed `StringUtilTest#assertExceptionThrown(...)` as well, so, should that be placed in another commit too?
> 
> It'll be helpful if you mention in your commit message that [7/19] will only be performing mechanical changes, while the subsequent commit(s) (e.g. [8/19]) will perform more complex changes. This helps reviewers like myself understand why you're splitting it up :)

Thanks! I've done so. I'm not sure why StringUtilTest#assertExceptionThrown got mixed together in [7/19]. I've taken it out into a separate commit. Thanks!

Zhiyuan-Amos commented on src/test/java/systemtests/ClockExtension.java:
```diff
@@ -54,4 +41,5 @@ public void setInjectedClockToCurrentTime() {
         injectedClock = Clock.fixed(Instant.now(), ZoneId.systemDefault());
         StatusBarFooter.setClock(injectedClock);
     }
+
```
unnecessary change :P

Zhiyuan-Amos commented on src/test/java/systemtests/AddressBookSystemTest.java:
```diff
@@ -48,6 +49,10 @@
     public static ClockRule clockRule = new ClockRule();
 
+    public static final Path SAVE_LOCATION_FOR_TESTING = TestUtil.getFilePathInSandboxFolder("sampleData.json");
+    protected static final Path DEFAULT_CONFIG_LOCATION_FOR_TESTING =
```
can these variables have `private` visibility?

Zhiyuan-Amos replied:
can these variables have `private` visibility?

Zhiyuan-Amos commented on src/test/java/seedu/address/ui/MainWindowCloseTest.java:
```diff
@@ -25,17 +26,20 @@
-    public final TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    @TempDir
+    public Path temporaryFolder;
```
can we not declare this as a variable? I saw this [example](https://stackoverflow.com/a/6185359/8828382): `writeItemsToFile(@TempDir Path tempDir)`

sijie123 replied:
Hmm.. I declared this as a variable mainly because I followed the structure of the previous TemporaryDirectory rule.

Yes, the way that you proposed would work. It's definitely possible (tests will all work) to migrate all tests to use the @TempDir Path tempDir syntax. However, every class that uses the tempDir needs the TempDir parameter in its method declaration, and nope, we can't abstract that away. This is because, say we have a
```
private String read(String file, @TempDir Path tempDir) {
        return tempDir.readFile();
}
```
We need to call this read method with `read("file.txt", tempDir);`. Every calling method still needs to provide the tempDir, else the code doesn't compile.

I've looked at JUnit 5's guides, there's no preference (from the developers) on which way should be adopted. So we can consider this way of writing temporary directories.

My only concern now is that we should maintain consistency. I think if we adopt this way of using temporary directories, then we should do so for every test that uses TempDirs.

I would think that the class-wide declaration is not that good a SE practice because the variable scope is larger, but it makes things a bit easier for readers because there's one less variable in the method declaration for all tests that require a tempdir. Classes like ConfigUtilTest have multiple tests that read from / write to a temporary directory, so I think it might be cleaner to just declare tempDir globally.

Zhiyuan-Amos commented on src/test/java/seedu/address/TestApp.java:
```diff
@@ -25,20 +24,15 @@
-    public TestApp() {
-    }
+    protected Path saveFileLocation = null;
+    protected Path configFileLocation = null;
```
is there a reason for initialising these variables to null? Also, their visibility can be set to `private`.

Zhiyuan-Amos commented on src/test/java/seedu/address/testutil/DirectoryInitUtil.java:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * Creates a temporary file with a randomized filename prefixed by {@code name} in the {@code tempDir}.
+     */
+    public static Path createTemporaryFileInFolder(Path tempDir, String name) throws IOException {
```
let's rename `name` to `prefix` since that's what `name` really is. rmb to update header comments too.

Zhiyuan-Amos commented on src/test/java/seedu/address/testutil/DirectoryInitUtil.java:
```diff
@@ -0,0 +1,57 @@
+        Path relativePath = relativizeToTempDir(filePath, sourceDir, tempDir);
+        try {
+            Files.copy(filePath, relativePath, StandardCopyOption.REPLACE_EXISTING);
+        } catch (IOException e) {
```
we should throw the exception up instead (as is the case of createTemporaryFileInFolder).

sijie123 replied:
If I throw the exception up, the lambda function in initializeTemporaryDirectory will not compile because the consumer lambda `forEach(filePath -> copyFilesToTempDir(filePath, sourceDir, tempDir));` cannot throw checked exceptions like IOException.
One way I can get around this is by rethrowing an AssertionError instead. I've seen our code do it before (e.g. in AddressBookSystemTest#assertSelectedCardChanged), but I'm not sure if this is a good practice.

Zhiyuan-Amos commented on src/test/java/seedu/address/testutil/DirectoryInitUtil.java:
```diff
@@ -0,0 +1,57 @@
+
+    /**
+     * Helper method to copy a file located at {@code filePath} to {@code tempDir}. Files copied maintain
+     * relative directory structure with reference to {@code sourceDir}.
```
this comment should be placed in `initializeTemporaryDirectory` instead. Also, you're saying that sourceDir's directory tree is mirrored to tempDir?

sijie123 replied:
Thanks. Yes, the second sentence makes more sense to be placed in initializeTemporaryDirectory.

Also, yes, I intend to create a utility that replicates/mirrors the `sourceDir` directory tree to tempDir. I will reword the comment slightly to make that clear.

Zhiyuan-Amos commented on src/test/java/seedu/address/commons/util/ConfigUtilTest.java:
```diff
@@ -10,20 +10,26 @@
+    @BeforeEach
+    public void setUp() throws IOException {
+        DirectoryInitUtil.initializeTemporaryDirectory(TEST_DATA_FOLDER, tempDir);
+
```
additional new line

Zhiyuan-Amos commented:
Well done, we're nearly there! I hope the next iteration will be sufficient 🎉

sijie123 commented:
> What do you think of changing `DirectoryInitUtil` to an Object-Oriented class?
> 
> ```java
> public class Foo {
>     @TempDir
>     private Path tempDir;
> 
>     public void copy(Path sourceDir) {...}
>     public void createTemporaryFile(String prefix) {...}
>     public void createTemporaryFile() {...}
>     public void getTemporaryDirectory() {...}
> }
> ```

This would be the ideal design and I actually started out with this in mind - to wrap around JUnit 5's TempDir and create a object oriented class supporting more features like automatic initialization and teardown. However, when I tried to do that, the TempDir didn't get initialised (I ended up with a null pointer).
I showed my preliminary implementation and asked the JUnit 5 developers over on [StackOverflow](https://stackoverflow.com/questions/54647577/how-to-use-tempdir-in-extensions/54651578#54651578), and the reason this happens is because extensions cannot "depend on one another".

In more recent development [here](https://github.com/junit-team/junit5/issues/1786), the JUnit team has decided that it would be meaningful to give other extensions the path of the TempDir folder via ExtensionContext.Store. With this feature, I could potentially implement the Object Oriented TempDir as suggested.
However, this feature is slated for JUnit 5.5 M2, which is still in development.

I would certainly like to implement this feature as a future enhancement.

Zhiyuan-Amos commented on src/test/java/seedu/address/testutil/DirectoryInitUtil.java:
```diff
@@ -0,0 +1,60 @@
+     * Initialise a temporary directory by copying all files from {@code sourceDir} to {@code tempDir}.
+     * Files copied will mirror the directory structure of {@code sourceDir}.
+     */
+    public static void initializeTemporaryDirectory(Path sourceDir, Path tempDir) throws IOException {
```
Following https://github.com/se-edu/addressbook-level4/pull/959/files#r268985535, we have to remove `throws IOException`

Zhiyuan-Amos replied:
Once this has been removed, the caller methods can have the `throws...` removed too.

Zhiyuan-Amos commented:
For `DirectoryInitUtil`, I saw this piece of code [online](https://javarevisited.blogspot.com/2016/02/how-to-copy-non-empty-directory-in-java.html) concerning mirroring directories:

```
public static void copyDirectory(File sourceDir, File targetDir)
 throws IOException {
        if (sourceDir.isDirectory()) {
            copyDirectoryRecursively(sourceDir, targetDir);
        } else {
            Files.copy(sourceDir.toPath(), targetDir.toPath());
        }
    }

private static void copyDirectoryRecursively(File source, File target)
 throws IOException {
        if (!target.exists()) {
            target.mkdir();
        }

        for (String child : source.list()) {
            copyDirectory(new File(source, child), new File(target, child));
        }
    }
}
```

We can try-catch the IOException, thereby being able to simplify the `source.list()` segment by using stream. What do you think?

Zhiyuan-Amos commented on src/test/java/seedu/address/commons/util/JsonUtilTest.java:
```diff
@@ -5,17 +5,26 @@
  */
 public class JsonUtilTest {
 
-    private static final Path SERIALIZATION_FILE = TestUtil.getFilePathInSandboxFolder("serialize.json");
```
is `"serialize.json"` from the Sandbox Folder actually being used now?

Zhiyuan-Amos commented:
I think the commit organization concerning `TemporaryFolder` can be further refined:

```
[v5 11/30] JUnit: Upgrade to JUnit 5.4.0
[v5 12/30] Temporary folder: Create helper class to manage temporary directory contents
... // Notice gap in commit numbers
[v5 21/30] TemporaryFolder: Migrate temporary file creation to DirectoryInitUtil#createTemporaryFileInFolder
[v5 22/30] TemporaryFolder: Copy test data files to temporary folder before running tests
[v5 23/30] TemporaryFolder: Remove redundant addToTestDataPathIfNotNull
[v5 24/30] TemporaryFolder: Migrate getFilePathInSandBoxFolder to Path#resolve
[v5 25/30] TestUtil: Remove getFilePathInSandboxFolder
[v5 26/30] JsonUtilTest: Rename SERIALIZATION_FILE to serializationFile
```

Can these commits be organisized sequentially? :)

Zhiyuan-Amos commented on src/test/java/systemtests/AddressBookSystemTest.java:
```diff
@@ -45,8 +46,14 @@
+    public static Path tempDir;
+
+    private static final String SAVE_FILENAME_FOR_TESTING = "sampleData.json";
+    private static final String CONFIG_LOCATION_FOR_TESTING = "pref_testing.json";
```
are "sampleData.json" and "pref_testing.json" in the sandbox folder still being used?

sijie123 replied:
Nope. The sandbox is actually created when tests are run for the very first time, and these files: sampleData, pref_testing, serialize.json, are all created by the test methods when the tests are first run. 
Actually the existing behaviour leaves these files behind in the sandbox, which can potentially affect the results of other tests (e.g. a test assumes the existence of sampleData.json due to a side-effect of another test, and reads from it when it's actually not existent). By using a temporary directory instead of the sandbox, we prevent tests from interfering with one another.

That said, I realise that I forgot to remove the sandbox entry from our .gitignore. Since we no longer use the sandbox, we don't need this entry anymore.

Zhiyuan-Amos commented:
I'm good with this, after you've made these changes and submitted another iteration, you can tag another developer for review :)

sijie123 commented:
> For `DirectoryInitUtil`, I saw this piece of code [online](https://javarevisited.blogspot.com/2016/02/how-to-copy-non-empty-directory-in-java.html) concerning mirroring directories:
> 
> ```
> public static void copyDirectory(File sourceDir, File targetDir)
>  throws IOException {
>         if (sourceDir.isDirectory()) {
>             copyDirectoryRecursively(sourceDir, targetDir);
>         } else {
>             Files.copy(sourceDir.toPath(), targetDir.toPath());
>         }
>     }
> 
> private static void copyDirectoryRecursively(File source, File target)
>  throws IOException {
>         if (!target.exists()) {
>             target.mkdir();
>         }
> 
>         for (String child : source.list()) {
>             copyDirectory(new File(source, child), new File(target, child));
>         }
>     }
> }
> ```
> We can try-catch the IOException, thereby being able to simplify the `source.list()` segment by using stream. What do you think?
> 
> Just a side note, if we have Apache Utils as a dependency, we can simply call a method called `Files.copy` or something haha.

Thanks @Zhiyuan-Amos. This method is a lot cleaner and easier to understand. I've adapted it for use in the upcoming iteration. `source.list()` returns an array, and it gets quite ugly if I want to turn it into a stream. So I decided to stick with the enhanced for-loop instead.

I thought about the Apache Commons. It's certainly a lot more convenient to copy a directory if we can use their IO library. However, I wonder if it justifies including the entire dependency just for one use in DirectoryInitUtil. I think it's probably easier to implement our own directory copying feature.


> I think the commit organization concerning `TemporaryFolder` can be further refined:
> 
> ```
> [v5 11/30] JUnit: Upgrade to JUnit 5.4.0
> [v5 12/30] Temporary folder: Create helper class to manage temporary directory contents
> ... // Notice gap in commit numbers
> [v5 21/30] TemporaryFolder: Migrate temporary file creation to DirectoryInitUtil#createTemporaryFileInFolder
> [v5 22/30] TemporaryFolder: Copy test data files to temporary folder before running tests
> [v5 23/30] TemporaryFolder: Remove redundant addToTestDataPathIfNotNull
> [v5 24/30] TemporaryFolder: Migrate getFilePathInSandBoxFolder to Path#resolve
> [v5 25/30] TestUtil: Remove getFilePathInSandboxFolder
> [v5 26/30] JsonUtilTest: Rename SERIALIZATION_FILE to serializationFile
> ```
> Can these commits be organisized sequentially? :)

Thanks! I forgot to reorder after reducing the scope of the commits previously.
Since changes to TestApp does not depend on TemporaryFolder anymore. I will move TestApp [15/30-19/30] away to the start of the whole PR. In fact, I think it's isolated enough now to even perhaps warrant its own PR.
I'll also move [12/30] to right before [22/30], and reorder some commits in between.

sijie123 commented:
@pyokagan and @fzdy1914 , if you're free, could you also help me take a look at the PR? Thank you!

pyokagan commented on src/test/java/systemtests/SystemTestSetupHelper.java:
```diff
@@ -21,10 +21,11 @@
      */
-    public TestApp setupApplication(Supplier<ReadOnlyAddressBook> addressBook, Path saveFileLocation) {
+    public TestApp setupApplication(Supplier<ReadOnlyAddressBook> addressBook, Path saveFileLocation,
+                                    Path configFileLocation) {
```
Indent only by 6 spaces.

pyokagan commented on src/test/java/systemtests/AddressBookSystemTest.java:
```diff
@@ -50,6 +50,8 @@
 
     private static final Path SAVE_LOCATION_FOR_TESTING = TestUtil.getFilePathInSandboxFolder("sampleData.json");
+    private static final Path CONFIG_LOCATION_FOR_TESTING =
+            TestUtil.getFilePathInSandboxFolder("pref_testing.json");
```
Can be one line

pyokagan commented on src/test/java/seedu/address/TestApp.java:
```diff
@@ -53,7 +51,7 @@ public TestApp(Supplier<ReadOnlyAddressBook> initialDataSupplier, Path saveFileL
     protected Config initConfig(Path configFilePath) {
         Config config = super.initConfig(configFilePath);
-        config.setUserPrefsFilePath(DEFAULT_PREF_FILE_LOCATION_FOR_TESTING);
+        config.setUserPrefsFilePath(configFileLocation);
```
This is the *preferences file*, not the config file. The field name is misnamed.

pyokagan commented on src/test/java/seedu/address/TestApp.java:
```diff
@@ -24,7 +24,7 @@
 public class TestApp extends MainApp {
 
-    private Supplier<ReadOnlyAddressBook> initialDataSupplier = () -> null;
+    private Supplier<ReadOnlyAddressBook> initialDataSupplier;
```
Can be made final as well.

sijie123 replied:
Hmm...? It's modified in the constructor...
```
public TestApp(Supplier<ReadOnlyAddressBook> initialDataSupplier, ...) {
  this.initialDataSupplier = initialDataSupplier;
  ...
```
That said, this.initialDataSupplier is assigned but not used. We could remove it entirely, I suppose.

pyokagan replied:
@sijie123 
> Hmm...? It's modified in the constructor so it can't be final:

Yes it can ;-)

Anyway, maybe let's not bother. It's tangential to the PR anyway.

pyokagan commented on src/test/java/seedu/address/testutil/Assert.java:
```diff
@@ -1,45 +1,34 @@
-        throw new AssertionError(String.format(
-                "Expected %s to be thrown, but nothing was thrown.", expectedException.getName()));
+    public static void assertThrows(Class<? extends Throwable> expectedType, String expectedMessage,
+                                    Executable executable) {
```
6 spaces indent.

pyokagan commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -104,57 +105,55 @@ public void getFilteredPersonList_modifyList_throwsUnsupportedOperationException
-    private void assertCommandSuccess(String inputCommand, String expectedMessage, Model expectedModel) {
-        assertCommandBehavior(null, inputCommand, expectedMessage, expectedModel);
+    private void assertCommandSuccess(String inputCommand, String expectedMessage,
+                                      Model expectedModel) throws Exception {
```
6 spaces indent.

pyokagan commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -104,57 +105,55 @@ public void getFilteredPersonList_modifyList_throwsUnsupportedOperationException
+     * Executes the command and confirms that
+     * - no exceptions are thrown
+     * - the feedback message is equal to {@code expectedMessage}.
+     * - the internal model manager state is the same as that in {@code expectedModel}.
```
Add `<br>` to list items.

pyokagan commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -104,57 +105,55 @@ public void getFilteredPersonList_modifyList_throwsUnsupportedOperationException
+                                      Model expectedModel) throws Exception {
+        CommandResult result = logic.execute(inputCommand);
+        assertEquals(expectedMessage, result.getFeedbackToUser());
+        assertEquals(model, expectedModel);
```
`assertEquals(expectedModel, model)`

pyokagan commented:
Sorry, please `s/6 spaces/8 spaces/g` everywhere. I can't count when half asleep.

damithc commented:
> Parameter lists on multiple lines uses 6-spaces indent, please do a sweep over all the commits to fix all such cases.

6 spaces or 8 spaces?

sijie123 commented:
> [3/33]
> 
> > "Furthermore, TestApp now needs to be initialised with the value of
> > configFileLocation **prefFileLocation**. The relevant method calls have been updated to
> > supply a prefFileLocation."

Nani I thought I did a find and replace... 😱

Zhiyuan-Amos commented:
btw, in the future, don't push till reviewers are done reviewing; now I can't use CanIHasReview to select the commits to review :P small thing though, just take note :)

Zhiyuan-Amos commented on src/test/java/seedu/address/testutil/DirectoryInitUtil.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    /**
+     * Creates a temporary file with a randomized filename in the {@code tempDir}
```
Missing period

Zhiyuan-Amos commented on src/test/java/seedu/address/testutil/DirectoryInitUtil.java:
```diff
@@ -0,0 +1,76 @@
+     * Initialise a temporary directory by copying all files from {@code sourceDir} to {@code tempDir}.
+     * Files copied will mirror the directory structure of {@code sourceDir}.
+     */
+    public static void initializeTemporaryDirectory(Path sourceDir, Path tempDir) {
```
header comment uses 's' for initialise, but here uses 'z' :p

Zhiyuan-Amos commented on src/test/java/systemtests/ClockExtension.java:
```diff
@@ -16,31 +16,18 @@
-
-    protected void after() {
+    @Override
+    public void afterEach(ExtensionContext context) throws Exception {
```
unnecessary `throws Exception`

Zhiyuan-Amos commented on src/test/java/systemtests/ClockExtension.java:
```diff
@@ -16,31 +16,18 @@
-                }
-            }
-        };
+    public void beforeEach(ExtensionContext context) throws Exception {
```
unnecessary `throws Exception`

sijie123 commented:
Hi @pyokagan and @Zhiyuan-Amos, can I get you guys to help me take a look at my latest CIHR?
I'm hoping to get it improved by the end of this week, so that we can merge it into AB4 soon. :)
Thanks!

Zhiyuan-Amos commented:
@sijie123

> Group assertions together using assertAll for related assertions (Not sure if we want to do this in this PR? Leaving it as separate asserts right now as feedback to user is clearer when tests fail)

From what I'm reading [here](https://stackoverflow.com/a/40800468/8828382):

> The interesting thing about assertAll is that it always checks all of the assertions that are passed to it, no matter how many fail. If all pass, all is fine - if at least one fails you get a detailed result of all that went wrong (and right for that matter).

I think it won't face the issue that you've mentioned. But don't worry, I don't think you need to resolve this in this PR :P

pyokagan commented on src/test/java/systemtests/AddressBookSystemTest.java:
```diff
@@ -93,6 +95,10 @@ protected Path getDataFileLocation() {
         return SAVE_LOCATION_FOR_TESTING;
     }
 
+    protected Path getPreferenceFileLocation() {
```
Shall we abbreviate this to `getPrefFileLocation()` as well to be consistent with the rest of the code?

pyokagan commented on src/test/java/seedu/address/logic/commands/CommandTestUtil.java:
```diff
@@ -115,18 +116,13 @@ public static void assertCommandFailure(Command command, Model actualModel, Comm
+        assertEquals(expectedFilteredList, actualModel.getFilteredPersonList());
+        assertEquals(expectedSelectedPerson, actualModel.getSelectedPerson());
+        assertEquals(expectedCommandHistory, actualCommandHistory);
 
```
We can remove the empty line here.

pyokagan commented on src/test/java/seedu/address/commons/util/StringUtilTest.java:
```diff
@@ -61,31 +56,25 @@ public void isNonZeroUnsignedInteger() {
-        errorMessage.ifPresent(message -> thrown.expectMessage(message));
-        StringUtil.containsWordIgnoreCase(sentence, word);
+        assertThrows(NullPointerException.class, ()
+            -> StringUtil.containsWordIgnoreCase("typical sentence", null));
```
This can be on the same line.

pyokagan commented on src/test/java/seedu/address/commons/core/index/IndexTest.java:
```diff
@@ -3,15 +3,18 @@
         // invalid index
-        assertCreateOneBasedFailure(0);
+        assertThrows(IndexOutOfBoundsException.class, () ->
+                Index.fromOneBased(0)
```
Any reason why this is on a separate line? Seems like it can be on the same line as the previous.

pyokagan commented on src/test/java/seedu/address/commons/core/index/IndexTest.java:
```diff
@@ -25,7 +28,9 @@ public void createOneBasedIndex() {
         // invalid index
-        assertCreateZeroBasedFailure(-1);
+        assertThrows(IndexOutOfBoundsException.class, () ->
+                Index.fromZeroBased(-1)
```
Same here.

pyokagan commented on src/test/java/seedu/address/commons/core/index/IndexTest.java:
```diff
@@ -3,15 +3,18 @@
+import static seedu.address.testutil.Assert.assertThrows;
 
-import org.junit.Test;
+import org.junit.jupiter.api.Test;
```
Is this intended in this commit? Doesn't seem so.

pyokagan commented on src/test/java/seedu/address/commons/util/StringUtilTest.java:
```diff
@@ -1,12 +1,12 @@
 import java.io.FileNotFoundException;
 
-import org.junit.Test;
+import org.junit.jupiter.api.Test;
```
Is this intended in this commit?

pyokagan commented on src/test/java/seedu/address/testutil/DirectoryInitUtil.java:
```diff
@@ -0,0 +1,76 @@
+     * Initialize a temporary directory by copying all files from {@code sourceDir} to {@code tempDir}.
+     * Files copied will mirror the directory structure of {@code sourceDir}.
+     */
+    public static void initializeTemporaryDirectory(Path sourceDir, Path tempDir) {
```
This is just a "copy directory tree" operation isn't it? Any reason why it needs such a specific `initializeTemporaryDirectory` name?

pyokagan replied:
Actually, looking down the commits, it seems that this may not even be needed?

pyokagan commented on src/test/java/seedu/address/testutil/DirectoryInitUtil.java:
```diff
@@ -0,0 +1,76 @@
+     * Helper method to copy a {@code source} file to a {@code target} file.
+     * If {@code source} is a directory, all files in {@code source} will be copied recursively to {@code target}.
+     */
+    private static void copyRecursively(File source, File target) {
```
We don't use the legacy `File` API in AB-4 (except for interfacing with legacy APIs). Please find a way to use the nio API.

pyokagan commented on src/test/java/seedu/address/testutil/DirectoryInitUtil.java:
```diff
@@ -0,0 +1,76 @@
+     * Helper method to recursively copy all contents of a {@code source} directory to a {@code target} directory.
+     * The {@code target} directory file structure will mirror that of the {@code source} directory.
+     */
+    private static void copyDirectoryRecursively(File source, File target) {
```
Seems like this method can be inlined into `copyRecursively()`.

pyokagan commented on src/test/java/seedu/address/testutil/DirectoryInitUtil.java:
```diff
@@ -0,0 +1,76 @@
+     * This is a wrapper method over Files#copy, to throw an AssertionError instead of an IOException so that
+     * calling methods do not need to repeatedly check for IOException.
+     */
+    private static void copyFile(File source, File target) {
```
This too.

pyokagan commented on src/test/java/seedu/address/testutil/DirectoryInitUtil.java:
```diff
@@ -0,0 +1,76 @@
+            Files.copy(source.toPath(), target.toPath());
+        } catch (IOException ioe) {
+            throw new AssertionError("Failed to copy file from " + source.getPath()
+                    + " to " + target.getPath() + ". Error: " + ioe.getMessage());
```
You can just chain exceptions using the last argument, like this:
```java
throw new AssertionError("Failed to copy file...", ioe)
```
and Java/IDEs will generate a nice stack trace that details all chained exceptions.

pyokagan commented on src/test/java/seedu/address/testutil/DirectoryInitUtil.java:
```diff
@@ -0,0 +1,76 @@
+    /**
+     * Creates a temporary file with a randomized filename prefixed by {@code prefix} in the {@code tempDir}.
+     */
+    public static Path createTemporaryFileInFolder(Path tempDir, String prefix) {
```
Seems like you can just shorten the name to `createTempFileInDir` or `createTempFile`.

Now, I also have my reservations about whether there is a *need* for creating files with random unique filenames in temporary directories, but I'll have to see the other commits to be sure...

....

I checked the rest of the commits, and I really do think `createTemporaryFileInFolder` is not needed at all.

pyokagan replied:
> Now, I also have my reservations about whether there is a need for creating files with random unique filenames in temporary directories, but I'll have to see the other commits to be sure...

So, the general reason why tests should usually not need to create files with random unique filenames in a temporary directory is because they *own* the temporary directory and have full control over it, and so a test would be able to choose its own filenames which are guaranteed to not conflict.

The only time where we need to have random unique filenames is when a directory is shared between multiple tests. This is not the case here though.

pyokagan commented on src/test/java/seedu/address/testutil/DirectoryInitUtil.java:
```diff
@@ -0,0 +1,76 @@
+/**
+ * A set of helper methods to manage temporary directories.
+ */
+public class DirectoryInitUtil {
```
The name seems inconsistent with the Javadoc, the Javadoc says "A set of helper methods to manage temporary directories.", but the name is "DirectoryInitUtil". I'd have expect the class name to contain some mention of temporary directories or vice versa.

pyokagan commented on src/test/java/seedu/address/testutil/DirectoryInitUtil.java:
```diff
@@ -0,0 +1,76 @@
+     */
+    private static void copyDirectoryRecursively(File source, File target) {
+        target.mkdir();
+        for (String child: source.list()) {
```
Need a space between the `:`

pyokagan commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -37,16 +37,18 @@
+    @BeforeEach
+    public void setUp() {
+        JsonAddressBookStorage addressBookStorage =
+                new JsonAddressBookStorage(createTemporaryFileInFolder(temporaryFolder));
```
Hrmph, who was the one that used `temporaryFolder.newFile()` even though it is not needed at all?

...

[Turns out it's me](https://github.com/se-edu/addressbook-level4/commit/89ec0382bcd3fe34edf1205ff02248ae5f954074), which is great, because I do think that we can just replace it with `temporaryFolder.resolve("addressbook.json")` or something so we don't need to use `createTemporaryFileInFolder()`.

pyokagan commented on src/test/java/seedu/address/storage/StorageManagerTest.java:
```diff
@@ -18,23 +18,20 @@
-        JsonAddressBookStorage addressBookStorage = new JsonAddressBookStorage(getTempFilePath("ab"));
-        JsonUserPrefsStorage userPrefsStorage = new JsonUserPrefsStorage(getTempFilePath("prefs"));
+        JsonAddressBookStorage addressBookStorage =
+                new JsonAddressBookStorage(createTemporaryFileInFolder(testFolder, "ab"));
```
The original code (`getTempFilePath()`) wasn't creating temporary files here, so it can just be `testFolder.resolve("ab")`.

sijie123 commented:
Cleaned up this PR by moving most tangential commits to other PRs.

For temporary folders, I agree that the existing implementation (leaving `TEST_DATA_FOLDER` as it is) would be less prone to error. Now that we don't need to populate a directory, and with no need to create a temporary file with a random name, there is no longer a need for DirectoryInitUtil.

I'm going to put the Sandbox folder issue aside for a bit to focus on making good the most important part: migration to JUnit 5.

> Okay, some playing around shows that ClockRule is not taking effect in
the system tests after this commit, so I wonder why the system tests are
passing... Maybe they have become flaky? In any case, it seems
dangerous.

ClockRule actually takes effect in system tests (at least in v9). The reason why, after [13/16] (i.e. switch to JUnit 5 Runner but still using ClockRule), ClockRule still works is because of the JUnit-Vintage Engine. We included it a long while back when we first wrote tests in JUnit 5, but this Vintage Engine is responsible for finding and running APIs written with JUnit 4.
"Just make sure that the junit-vintage-engine artifact is in your test runtime path. In that case JUnit 3 and JUnit 4 tests will automatically be picked up by the JUnit Platform launcher." [JUnit 5 User Guide](https://junit.org/junit5/docs/current/user-guide/#migrating-from-junit4-running)

Please also ignore [1/16-4/16], as these commits are already in a separate PR. I'm leaving it here so that I don't have to keep resolving merge conflicts when these PRs get merged; I just need to rebase later.

pyokagan commented:
@sijie123 I need to go out for a bit. In the meantime could you rebase and send another CIHR?

> ClockRule actually takes effect in system tests (at least in v9). The reason why, after [13/16] (i.e. switch to JUnit 5 Runner but still using ClockRule), ClockRule still works is because of the JUnit-Vintage Engine.

I haven't tested v9, but in the previous revision when I made that comment I added a `throws RuntimeException()` to ClockRule. It didn't fire.

sijie123 commented:
I did a quick rebase + CIHR, not yet resolved with your comments in #1009.
I'll do that later tonight.

pyokagan commented on src/test/java/seedu/address/testutil/Assert.java:
```diff
@@ -1,53 +1,33 @@
-     * Asserts that the {@code callable} throws the {@code expected} Exception.
+     * Asserts that the {@code executable} throws the {@code expectedType} Exception.
+     * This is a wrapper method that invokes
+     * {@code Assertions.assertThrows(Class<? extends Throwable>, Executable)}, to maintain consistency
```
Shouldn't this be a `@link`?
```
{@link Assertions#assertThrows(Class, Executable)}
```

pyokagan commented on src/test/java/seedu/address/testutil/Assert.java:
```diff
@@ -1,53 +1,33 @@
+     * Asserts that the {@code executable} throws the {@code expectedType} Exception.
+     * This is a wrapper method that invokes
+     * {@code Assertions.assertThrows(Class<? extends Throwable>, Executable)}, to maintain consistency
+     * with our custom {@see assertThrows(Class<? extends Throwable>, String, Executable)} method.
```
Shouldn't this be a `@link`?

```
{@link #assertThrows(Class, String, Executable)}
```

Remember to rewrap the javadoc as required.

pyokagan commented on src/test/java/seedu/address/testutil/Assert.java:
```diff
@@ -1,53 +1,33 @@
+     * This is a wrapper method that invokes
+     * {@code Assertions.assertThrows(Class<? extends Throwable>, Executable)}, to maintain consistency
+     * with our custom {@see assertThrows(Class<? extends Throwable>, String, Executable)} method.
+     * To standardize API calls in this project, users should use this method instead of JUnit#assertThrows.
```
We can't use the `#` notation this way. Either `JUnit's assertThrows` or use `@link` with the full method signature.

pyokagan commented on src/test/java/seedu/address/testutil/Assert.java:
```diff
@@ -1,53 +1,33 @@
-     * {@code assertThrows(Class<? extends Throwable>, VoidCallable)} instead.
-     * {@see assertThrows(Class<? extends Throwable>, VoidCallable}
+     * Asserts that the {@code executable} throws the {@code expectedType} Exception with the
+     * {@code expectedMessage}. If there's no need for the verification of the exception's error
```
The first sentence (up to `{@code expectedMessage}`) can be on one line. This is especially important because the first sentence is important (it is the javadoc "summary" sentence), and should be kept short and on one line as much as possible.

pyokagan commented on src/test/java/seedu/address/testutil/Assert.java:
```diff
@@ -1,53 +1,33 @@
-     * {@see assertThrows(Class<? extends Throwable>, VoidCallable}
+     * Asserts that the {@code executable} throws the {@code expectedType} Exception with the
+     * {@code expectedMessage}. If there's no need for the verification of the exception's error
+     * message, call {@code assertThrows(Class<? extends Throwable>, Executable)} instead.
```
Shouldn't this use `@link`?

```
{@link #assertThrows(Class, Executable)}
```

pyokagan commented on src/test/java/seedu/address/testutil/Assert.java:
```diff
@@ -1,53 +1,33 @@
+     * Asserts that the {@code executable} throws the {@code expectedType} Exception with the
+     * {@code expectedMessage}. If there's no need for the verification of the exception's error
+     * message, call {@code assertThrows(Class<? extends Throwable>, Executable)} instead.
+     * {@see assertThrows(Class<? extends Throwable>, Executable)}
```
This should be a `@see` on its own, separated by one blank line:

```java
/**
 * Call ... instead.
 *
 * @see #assertThrows(Class, Executable)
 */
```
(Note that the method reference syntax you used is wrong)

pyokagan commented on build.gradle:
```diff
@@ -50,7 +50,7 @@ dependencies {
     testImplementation group: 'junit', name: 'junit', version: '4.12'
     testImplementation group: 'org.testfx', name: 'testfx-core', version: testFxVersion
-    testImplementation group: 'org.testfx', name: 'testfx-junit', version: testFxVersion
+    testImplementation group: 'org.testfx', name: 'testfx-junit5', version: testFxVersion
```
Uh, why was this change made? Commit message doesn't say.

pyokagan replied:
Seems like this library must be present once we start migrating tests (GUI tests especially) to use the JUnit5 runtime. *However*, the `testfx-junit` library must remain until we finish migrating *all* tests.

pyokagan commented on src/test/java/seedu/address/ui/GuiUnitTest.java:
```diff
@@ -2,19 +2,23 @@
 /**
  * A GUI unit test class for AddressBook.
  */
+@ExtendWith(StageExtension.class)
```
This seems quite arbitrary... can't we just do `UiPartExtension extends StageExtension`?

sijie123 replied:
Weird... I cannot remember why I used this `@ExtendWith` method... And now I don't see any benefits of doing so either.

`UiPartExtension extends StageExtension` as you suggested works correctly, and it actually follows our previous implementation more closely. So yes, I will change to use `UiPartExtension extends StageExtension`.

pyokagan commented on src/test/java/seedu/address/ui/testutil/StageExtension.java:
```diff
@@ -0,0 +1,22 @@
+ * Properly sets up and tears down a JavaFx stage for our testing purposes.
+ */
+public class StageExtension implements BeforeEachCallback, AfterEachCallback {
+
```
Remove blank line here for symmetry.

pyokagan commented on src/test/java/seedu/address/ui/GuiUnitTest.java:
```diff
@@ -2,19 +2,23 @@
 public abstract class GuiUnitTest {
-    @Rule
-    public final UiPartRule uiPartRule = new UiPartRule();
+
```
Remove blank line here for symmetry.

pyokagan commented on src/test/java/systemtests/AddressBookSystemTest.java:
```diff
@@ -264,11 +264,11 @@ protected void assertStatusBarUnchanged() {
     /**
      * Asserts that only the sync status in the status bar was changed to the timing of
-     * {@code ClockRule#getInjectedClock()}, while the save location remains the same.
+     * {@code ClockExtension#getInjectedClock()}, while the save location remains the same.
```
I wonder if this should be changed to `@link`

pyokagan commented:
btw, [5/13] is done too early. We can only use JUnit5's `assumeFalse()` once the test in question has been migrated to use the JUnit5 test runner. Otherwise instead of the test skipping we get this test failure:

```
seedu.address.ui.HelpWindowTest > focus_helpWindowNotFocused_focused FAILED
    org.opentest4j.TestAbortedException: Assumption failed: Test skipped in headless mode: Window focus behavior is buggy.
        at org.junit.jupiter.api.Assumptions.throwTestAbortedException(Assumptions.java:249)
        at org.junit.jupiter.api.Assumptions.assumeFalse(Assumptions.java:184)
        at seedu.address.ui.HelpWindowTest.focus_helpWindowNotFocused_focused(HelpWindowTest.java:59)
```

 Since the test is only migrated to use the JUnit5 test runner in [10/13], [5/13] should be squashed into [5/13].

sijie123 commented:
* Updated with comments from #1009 
* Squashed the assume into a later commit when HelpWindowTest migrates to JUnit 5 API
* Moved SystemTests to a later commit together with ClockExtension
* Resolved the `@link` issues (I hope I've caught them all)

I wonder if the current organisation is the easiest to understand, because, for example, I now need to squash the SystemTests together with the ClockExtension commit. I can't put it afterwards, because the JUnit 4 runner won't pick up the ClockExtension. I can put it before, but it's kinda hackish and works only because we have the JUnit Vintage runner that picks up and processes the ClassRule. So it's technically not correct to put it before either...

Zhiyuan-Amos commented on src/test/java/seedu/address/testutil/Assert.java:
```diff
@@ -1,53 +1,34 @@
-     * {@see assertThrows(Class<? extends Throwable>, VoidCallable}
+     * {@link #assertThrows(Class, Executable)} instead.
+     *
+     * {@see #assertThrows(Class, Executable)}
```
should just be `@see #assertThrows(Class, Executable)` i.e. no curly braces :P

pyokagan commented:
@sijie123 I didn't do a full review because I need to go off now, but my comments in [6/10] may be interesting, you can submit a new iteration if needed before I get back.

sijie123 commented:
> Something doesn't make sense here... The cited webpage shows that we
> only need the testfx-junit5 library if we are using its
> `org.testfx.framework.junit5.ApplicationExtension`, aka we do:
> 
> ```java
> @ExtendWith(ApplicationExtension.class)
> ```
> 
> But we don't use that anywhere in our codebase...
> 
> So, perhaps we don't even need the `testfx-junit5` library?
> In fact, after I removed it, the tests seem to run as normal... (Maybe I
> conflated it with the "systemtests migrated to the junit5 test runner
> too early" issue in v10).
> 
> Along the same train of thought, maybe we didn't even need the
> `testfx-junit` library in the first place, because we didn't even use
> its `ApplicationTest` either?
> 
> _(tries it out on master)_
> 
> Yup, it seems we don't even need the `testfx-junit` library in the first
> place.

That's interesting. I've tested it on master too, and yes, the tests actually work without needing testfx-junit.
I've dug around to understand why this is the case - and I believe it's mainly because our StageRule/StageExtension calls registerPrimaryStage(). This is exactly what [testfx-junit's ApplicationTest](https://github.com/TestFX/TestFX/blob/master/subprojects/testfx-junit/src/main/java/org/testfx/framework/junit/ApplicationTest.java) does.
Similarly, during system tests, `SystemTestSetupHelper` sets up the UI environment.

So for now, it means that we won't need testfx-junit*. In that case, in [v11 6/10], I won't make the migration to testfx-junit5 anymore. However, I don't think it's appropriate to remove testfx-junit in this PR, so maybe I'll open a separate PR in the future (after my paper on Tuesday maybe (?)) to document and remove it from our build.gradle.

pyokagan commented:
@sijie123 
> However, I don't think it's appropriate to remove testfx-junit in this PR, so maybe I'll open a separate PR in the future (after my paper on Tuesday maybe (?)) to document and remove it from our build.gradle.

I think it should be done as a preparatory PR or in this PR itself. It would not be correct if, after a PR called "Migrate from JUnit 4 to JUnit 5" is merged, a JUnit4 dependency still remains.

pyokagan commented on src/test/java/seedu/address/storage/JsonAddressBookStorageTest.java:
```diff
@@ -23,9 +22,9 @@
 
-    @Rule
-    public TemporaryFolder testFolder = new TemporaryFolder();
 
```
Remove this extra blank line

pyokagan commented on src/test/java/seedu/address/storage/JsonUserPrefsStorageTest.java:
```diff
@@ -21,9 +20,9 @@
 
-    @Rule
-    public TemporaryFolder testFolder = new TemporaryFolder();
 
```
Remove this extra blank line.

pyokagan commented on src/test/java/seedu/address/ui/MainWindowCloseTest.java:
```diff
@@ -25,17 +25,19 @@
 public class MainWindowCloseTest extends GuiUnitTest {
-    @Rule
-    public final TemporaryFolder temporaryFolder = new TemporaryFolder();
+
```
Remove this blank line for symmetry.

pyokagan commented on src/test/java/seedu/address/commons/util/ConfigUtilTest.java:
```diff
@@ -22,8 +21,8 @@
-    public TemporaryFolder testFolder = new TemporaryFolder();
+    @TempDir
+    public Path tempDir;
 
```
Hrmph, I noticed that 5cabb37a1ec20e59d2bb94 introduced an extra blank line here. Perhaps we can sneakily remove it in this commit as well.

sijie123 commented:
After this PR, there's still 3 more places (StringUtilTest, ConfigUtilTest, ConfigTest) where there are still empty lines. I couldn't find a good place to put in this PR, as I didn't really touch the code around the blank lines in my commits. Maybe after this PR, I can do a small PR to fix these blank lines separately.
